src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+// Microsoft Serial Mouse emulation originally wrtitten by Jonathan Campbell
+// Wheel, Logitech, and Mouse Systems mice added by Roman Standzikowski (FeralChild64)
+
+// Reference:
+// - https://roborooter.com/post/serial-mice
+// - https://www.cpcwiki.eu/index.php/Serial_RS232_Mouse
+
+#include "serialmouse.h"
+
+#include "checks.h"
+#include "mouse.h"
+
+CHECK_NARROWING();
+
+
+CSerialMouse::CSerialMouse(uint8_t id, CommandLine* cmd): CSerial(id, cmd),
+    port_num(static_cast<uint16_t>(id + 1)), 
+    config_type(MouseType::NO_MOUSE),
+    config_auto(false),
+    mouse_type(MouseType::NO_MOUSE),
+    mouse_bytelen(0),
+    mouse_has_3rd_button(false),
+    mouse_has_wheel(false),
+    mouse_port_valid(false),
+    smooth_div(1),
+    send_ack(true),
+    packet_len(0),
+    xmit_idx(0xff),
+    xmit_2part(false),
+    xmit_another_move(false),
+    xmit_another_button(false),
+    mouse_buttons(0),
+    mouse_delta_x(0),
+    mouse_delta_y(0),
+    mouse_delta_w(0)
+{
+    std::string type_string;
+    bool use_default = !cmd->FindStringBegin("type:", type_string, false);
+
+    if (type_string == "2btn") {
+        config_type = MouseType::MICROSOFT;
+        config_auto = false;
+    } else if (type_string == "2btn+msm") {
+        config_type = MouseType::MICROSOFT;
+        config_auto = true;
+    } else if (type_string == "3btn") {
+        config_type = MouseType::LOGITECH;
+        config_auto = false;
+    } else if (type_string == "3btn+msm") {
+        config_type = MouseType::LOGITECH;
+        config_auto = true;
+    } else if (type_string == "wheel") {
+        config_type = MouseType::WHEEL;
+        config_auto = false;
+    } else if (type_string == "wheel+msm" || use_default) {
+        config_type = MouseType::WHEEL;
+        config_auto = true;
+    } else if (type_string == "msm") {
+        config_type = MouseType::MOUSE_SYSTEMS;
+        config_auto = false;
+    } else {
+        LOG_ERR("MOUSE (COM%d): Invalid type '%s'", port_num, type_string.c_str());
+        return;
+    }
+
+    std::string rate_string;
+    use_default = !cmd->FindStringBegin("rate:", rate_string, false);
+
+    if (rate_string == "smooth" || use_default)
+        // This is roughly equivalent of PS/2 mouse with sampling rate between 120 and 250 Hz,
+        // depends on mouse protocol, concrete events (reporting middle button state change or
+        // wheel movement needs more bandwidth for Logitech and wheel mice), and sometimes even
+        // driver sophistication (for Mouse Systems mouse a badly written driver can waste half
+        // of the movement sampling rate)
+        smooth_div = 5;
+    else if (rate_string != "normal") {
+        LOG_ERR("MOUSE (COM%d): Invalid rate '%s'", port_num, rate_string.c_str());
+        return;
+    }
```

Comment:
Reading from the config doesn't belong to the mouse emulation initialisation itself; it would be better to separate the config reading out, and just pass the parsed parameters to the constructor like other recently refactored modules are doing (e.g. `gus.cpp` and `gameblaster.cpp`).

Reply:
The problem is, that (contrary to GUS or GameBlaster) we can have several instances of CSerialMouse, for each COM port.
What you propose, in order to make sense, would need to be implemented somewhere in the general COM port support code, and refactoring should cover all the emulated serial port devices. Currently `nullmodem.cpp`, `directserial.cpp`, `softmodem.cpp` - all follow the same convention, they read their configuration options in the constructor.

IMHO this is not the right time to do such a rework.

Reply:
Sound reasonable 👍🏻